### PR TITLE
Redesign command line interface

### LINKED ISSUE #21: Should `benchcab` have only one main program entry point?
Currently, `benchcab` uses two main program entry points: `benchcab` and `benchsiterun`.
- `benchcab` does the setup to run all the different test configurations
- `benchsiterun` runs CABLE over the different test configurations and is invoked via the job script (which is submitted by `benchcab`)

Problem:
- ideally, `benchsiterun` is invoked only by `benchcab` through the job script where the necessary setup steps have been completed (for example creating the directory structures, checking out and building CABLE, etc). However, `benchsiterun` is accessible to the user at the same level as `benchcab` which may give the user the impression that it is an additional tool to `benchcab`. This forces us to include a number of checks in [benchsiterun.py ](https://github.com/CABLE-LSM/benchcab/blob/master/benchcab/benchsiterun.py) as we cannot trust the user to use `benchsiterun` correctly (for example checking the directory structure is valid, relevant input files and executables exist, etc).

Is there a better solution where `benchsiterun` can only be invoked internally by `benchcab` and not by the user?

A possible solution could be:

1. Adding a flag to `benchcab` that disables the job script submission and instead lets users run CABLE on the same compute node they happen to run `benchcab` from.
2. Implement the ability to run specific steps of `benchcab` in isolation. For example, the checkout step, the build step, and the run step.

In the job script, instead of running `benchsiterun`, we can run `benchcab` with a flag that specifies the run step to be executed (preventing any the job script related code from being executed).

### PR DESCRIPTION
**Pull request by SeanBryan51**
Currently, `benchcab` uses two main program entry points: `benchcab` and `benchsiterun`. Having `benchsiterun` accessible at the same level as `benchcab` allows the user to run `benchsiterun` incorrectly, for example if the user tries to run `benchsiterun` before the work directory is generated. Ideally `benchsiterun` is invoked only by the job script submitted by `benchcab` and should be "hidden" to the user.

Move the `benchsiterun` main program entry point into `benchcab`.

Add the ability to run specific steps in the workflow in isolation via [subcommands].

These steps include checkout, build, setting up the fluxnet task work directory and executing the fluxnet tasks.

Add the `--no-submit` flag to disable the job script submission when executing fluxnet tasks.

This flag can also be specified in the main workflow if the user wishes to execute tasks on the login node (if the computation is inexpensive enough so that the job doesn't get killed).

Add unit tests to test the command line arguments are parsed correctly.

This should provide a command line interface that makes it difficult for the user to incorrectly run `benchsiterun` (now `benchcab fluxnet-run-tasks --no-submit`).

Fixes #21

[subcommands]: https://docs.python.org/3/library/argparse.html#sub-commands

## Testing
- `pytest` success
- `pip install --user .` success
- `benchcab fluxnet` success
- `benchcab fluxnet --no-submit` success (ran with a single science configuration and AU-Tum experiment)

### CODE REVIEW

**Review by ccarouge — CHANGES_REQUESTED** [review]
I'm happy with sub-commands to do specific parts of the workflow. But I'd like to keep a default workflow that runs with a simple `benchcab` call. For me, the default would be to run the flux site case and the spatial case (once it exists). I.e. the default is to run a full workflow.

It would be great if we could get this behaviour back.

**Comment by SeanBryan51** [comment]
@ccarouge I propose instead of having `benchcab`, we have `benchcab run` for executing the default workflow?
This way, subcommands such as `checkout` or `build` won't inherit optional arguments that are not needed from the default `benchcab` command. For example, passing the `--no-submit` flag to `benchcab` is valid but this design allows passing the same flag to `benchcab checkout` which doesn't really make sense.

 I think sub commands is one of things where if you start using them, everything needs to become a sub command 😅

**Comment by ccarouge** [comment]
Yes, `benchcab run` is ok. That's easy enough to remember.

File: benchcab/cli.py [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+        add_help=False
+    )
+
+    # subcommand: 'benchcab fluxnet-setup-work-dir'
+    subparsers.add_parser(
+        'fluxnet-setup-work-dir',
+        parents=[args_help, args_subcommand],
+        help="Run the work directory setup step of the main fluxnet command.",
+        description="""Generates the benchcab work directory in the current working directory so
+        that tasks can be run.""",
```
Comment (ccarouge) [inline benchcab/cli.py:101]:
```suggestion
        help="Run the work directory setup step of the fluxnet command.",
        description="""Generates the benchcab site/run directory tree in the current working directory so
        that tasks can be run.""",
```

File: benchcab/cli.py [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+    )
+
+    # subcommand: 'benchcab fluxnet-run-tasks'
+    subparsers.add_parser(
+        'fluxnet-run-tasks',
+        parents=[args_help, args_subcommand, args_run_subcommand],
+        help="Run the fluxnet tasks of the main fluxnet command.",
+        description="""Runs the fluxnet tasks for the main fluxnet test suite. By default, this
+        command generates a PBS job script and submits it to the queue.""",
+        add_help=False
```
Comment (ccarouge) [inline benchcab/cli.py:112]:
```suggestion
        help="Run the fluxnet tasks of the fluxnet command.",
        description="""Runs the fluxnet tasks for the fluxnet test suite. By default, this
        command generates a PBS job script and submits it to the queue.""",
        add_help=False
```

**Review by ccarouge — APPROVED** [review]
Some minor modifications to fix but this can be merged after the fixes are applied to your liking.

File: benchcab/benchcab.py [hunk 3]
```diff
@@ -5,133 +5,126 @@
...
+    validate_environment(project=config['project'], modules=config['modules'])
 
-if __name__ == "__main__":
+    if args.subcommand == 'run':
+        benchcab_fluxnet(args, config, tasks=get_fluxnet_tasks(
+            realisations=config["realisations"],
+            science_config=config['science_configurations'],
+            met_sites=get_met_sites(config['experiment'])
+        ))
+        benchcab_spatial()
```
[the remark this answers is not in the document]
Comment (ccarouge) [reply benchcab/benchcab.py:96]: That's for the case `if args.subcommand == 'run':`

**Comment by SeanBryan51** [comment]
@ccarouge I've made the changes you requested. Can you review the changes made in [b8a9471](https://github.com/CABLE-LSM/benchcab/pull/59/commits/b8a9471442f1b2df1a4af60e9fb087e78debf967)? This change is addressing your comment: https://github.com/CABLE-LSM/benchcab/pull/59#discussion_r1168201303
> Would it be worth getting the list of tasks only once? I don't like having the call to `get_fluxnet_tasks` copied so many times. It's an easy line to mess up.
> 
> Maybe worth putting it after the call to `validate_environment`.

I thought putting `get_fluxnet_tasks` after the `validate_environment` line is not ideal as we would be calling it for every subcommand of benchcab. I've instead abstracted `get_fluxnet_tasks` into a single function `_initialise_tasks` which removes the repetition.

As for why I went with a class implementation: this was to reduce the repetition of common arguments between all the subcommands. This is all just code aesthetic at this point.

**Review by ccarouge — APPROVED** [review]
I think we can leave it at this for now. I think the final design might take a bit of discussion and it's not a priority now.

**Pull request merged**